controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+
+	mw, err := a.findManifestWork(mwName, homeCluster)
+	if err != nil && !errors.IsNotFound(err) {
+		a.log.Error(err, "failed to find ManifestWork")
+
+		return nil, err
 	}
 
-	switch action {
-	case rmn.ActionFailover:
-		return a.validateFailover(subscription, newHomeCluster)
-	case rmn.ActionFailback:
-		return a.validateFailback(subscription, newHomeCluster)
-	default:
-		return newHomeCluster, fmt.Errorf("unknown action %s", action)
+	if mw == nil {
+		a.log.Info(fmt.Sprintf("Manifestwork %s does not exist for cluster %s", mwName, homeCluster))
+
+		return nil, nil
 	}
+
+	a.log.Info(fmt.Sprintf("Manifestwork %s exists (%v)", mw.Name, mw))
+
+	return mw, nil
 }
 
-func (a *AVRInstance) validateFailover(
-	subscription *subv1.Subscription, newHomeCluster string) (string, error) {
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		switch newHomeCluster {
-		case d.PeerCluster:
-			return newHomeCluster, nil
-		case d.HomeCluster:
-			return newHomeCluster, errorswrapper.Wrap(ErrSameHomeCluster, newHomeCluster)
-		case d.PrevHomeCluster:
-			return newHomeCluster,
-				fmt.Errorf("miconfiguration detected on failover! (n:%s,p:%s)", newHomeCluster, d.PrevHomeCluster)
+func (a *AVRInstance) findManifestWork(mwName, homeCluster string) (*ocmworkv1.ManifestWork, error) {
+	if homeCluster != "" {
+		mw := &ocmworkv1.ManifestWork{}
+
+		err := a.reconciler.Get(a.ctx, types.NamespacedName{Name: mwName, Namespace: homeCluster}, mw)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return nil, fmt.Errorf("MW not found (%w)", err)
+			}
+
+			return nil, fmt.Errorf("failed to retrieve manifestwork (%w)", err)
 		}
+
+		return mw, nil
+	}
+
+	return nil, nil
+}
+
+func (a *AVRInstance) isVRGPrimary(mw *ocmworkv1.ManifestWork) (bool, error) {
+	vrgClientManifest := &mw.Spec.Workload.Manifests[0]
+	vrg := &rmn.VolumeReplicationGroup{}
+
+	err := yaml.Unmarshal(vrgClientManifest.RawExtension.Raw, &vrg)
+	if err != nil {
+		return false, errorswrapper.Wrap(err, fmt.Sprintf("unable to get vrg from ManifestWork %s", mw.Name))
 	}
 
-	return newHomeCluster, fmt.Errorf("unknown error (n:%s,avrStatus:%v)", newHomeCluster, a.instance.Status)
+	return (vrg.Spec.ReplicationState == volrep.Primary), nil
 }
 
-func (a *AVRInstance) validateFailback(
-	subscription *subv1.Subscription, newHomeCluster string) (string, error) {
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		switch newHomeCluster {
-		case d.PrevHomeCluster:
-			return newHomeCluster, nil
-		case d.HomeCluster:
-			return newHomeCluster, errorswrapper.Wrap(ErrSameHomeCluster, newHomeCluster)
-		case d.PeerCluster:
-			return "", fmt.Errorf("miconfiguration detected on failover! (n:%s,p:%s)", newHomeCluster, d.PrevHomeCluster)
+func (a *AVRInstance) ensureCleanup() error {
+	a.log.Info("ensure cleanup ", "cluster", a.instance.Status.ClusterToClean)
+
+	if a.instance.Status.ClusterToClean != "" {
```

Comment:
(continuation of clusterToClean status field comment) For example here, we can determine clusterToClean based on DRPeers and non-current cluster to check and ensure are clean.

If we fail to update status (post all actions are done) then, we will not store the cluster to clean and hence not clean it. May hence require checking ASoW (actual-state-of-the-world) instead of saving it in status.

Reply:
I agree with that. I'll change it.